src/System.Management.Automation/engine/lang/codegen.cs

```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+                value;
+        }
+
+        /// <summary>
+        /// Quote an argument, if needed, or if specifically requested, escaping characters accordingly,
+        /// handling escaping of wildcard patterns when argument is not already taken literally.
+        /// For example: QuoteArgument(userContent, quoteInUse, isLiteralArgument)
+        /// </summary>
+        /// <param name="value">The content to be used as an argument value taken literally.</param>
+        /// <param name="quoteInUse">The character to be quoted with.</param>
+        /// <param name="isLiteralArgument">Treat the argument as taken literally, wildcard escaping not required.</param>
+        /// <returns>Content quoted and escaped if required for use as an argument value.</returns>
+        public static string QuoteArgument(string value, char quoteInUse, bool isLiteralArgument)
+        {
+            return string.IsNullOrEmpty(value) ?
+                string.Empty :
+                QuoteArgument(isLiteralArgument ? value : WildcardPattern.Escape(value), quoteInUse);
+        }
+
+        /// <summary>
+        /// Quote an argument, if needed, or if specifically requested, escaping characters accordingly
+        /// For example: QuoteArgument(userContent, quoteInUse)
+        /// </summary>
+        /// <param name="value">The content to be used as an argument value taken literally.</param>
+        /// <param name="quoteInUse">The character to be quoted with.</param>
+        /// <returns>Content quoted and escaped if required for use as an argument value.</returns>
+        public static string QuoteArgument(string value, char quoteInUse)
+        {
+            if (string.IsNullOrEmpty(value))
+                return string.Empty;
+
+            char quoteToUse = quoteInUse;
+            if (quoteToUse == 0)
+                if (ShouldArgumentNotBeBareword(value))
+                    // argument value not compatible with bareword
+                    quoteToUse = '\'';
+                else
+                    // return unmodified argument value
+                    return value;
+
+            // quote and escape argument as needed
+            return quoteToUse + (quoteToUse.IsDoubleQuote() ? EscapeDoubleQuotedStringContent(value) :
+                EscapeSingleQuotedStringContent(value)) + quoteToUse;
+        }
+
+        // brute force method to determine when an argument's value cannot be used bareword
+        private static bool ShouldArgumentNotBeBareword(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+                return false;
+
+            // Rules for bareword arguments:
+            //   characters that cannot be in the first position: '@','#','<','>'
+            //   Pattern that cannot be in the first position: /[1-6]>/
+            bool requiresQuote = "@#<>".Contains(value[0]) ||
+                value.Length > 1 && '1' <= value[0] && value[0] <= '6' && value[1] == '>';
```

Comment:
Please use the pattern:
(uint)(value[0] - '1') <= '6' - '1'

Why do we need last "value[1] == '>'"? It is second position. Or previous comment is wrong?

Reply:
> Please use the pattern:
> (uint)(value[0] - '1') <= '6' - '1'
Why `unit`?  Does subtraction of [char]'s automatically become unit?

Reply:
I confirmed that it appears that compiler must upscale during arithmetic, so even though it looks like an upscale, its actually leaving it at the scale it performed it at.

Also, it looks like compiler does not pre-compute literal arithmetic (at least not in debug), so using `5` in place of `'6' - '1'` is another >10% faster yet.

Reply:
> at least not in debug

You should check release code to get all optimizations. If after that you still a bad code you should report in CoreClr repository because the pattern is used in CoreFX a lot.

> Maybe comment should read

Your comments is still not correct. 

>             //   characters that cannot be in the first position: '@','#','<','>' 
            //   Pattern that cannot be in the first position: /[1-6]>/

It seems it should be:
```
            //   characters that cannot be in the first position: '@','#','<','>' 
            //   Pattern that cannot be in the second position: /[1-6]>/
```

Reply:
Note that this code is outdated and was changed greatly, both the code and the comment referenced above had to change because of more patterns that needed to be checked.

I think there may still be some confusion, the pattern `/[1-6]>/` is two characters long, and it must NOT start in the first position of the 'value', (or value[0] and value[1] respectively).